sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -169,25 +184,56 @@ public Mono<TValue> value() {
-            Mono<TValue> valueMono = this.value.get();
+        public Mono<TValue> getOrCreateBackgroundRefreshTaskAsync(Function<TValue, Mono<TValue>> createRefreshFunction) {
+            if (this.refreshInProgress.compareAndSet(null, this.createBackgroundRefreshTask(createRefreshFunction))) {
+                logger.debug("Started a new background task");
+            } else {
+                logger.debug("Background refresh task is already in progress");
+            }
 
-            return valueMono.flatMap(value -> {
-                if(this.refreshInProgressCompleted.compareAndSet(false, true)) {
-                    this.refreshInProgress = createRefreshFunction.apply(value).cache();
-                    return this.refreshInProgress
+            Mono<TValue> refreshInProgressSnapshot = this.refreshInProgress.get();
+            return refreshInProgressSnapshot == null ? this.value.get() : refreshInProgressSnapshot;
+        }
+
+        private Mono<TValue> createBackgroundRefreshTask(Function<TValue, Mono<TValue>> createRefreshFunction) {
+            return this.value
+                        .get()
+                        .flatMap(cachedValue -> createRefreshFunction.apply(cachedValue))
                         .flatMap(response -> {
-                            this.value.set(Mono.just(response));
-                            this.refreshInProgressCompleted.set(false);
-                            return this.value.get();
-                        }).doOnError(e -> this.refreshInProgressCompleted.set(false));
+                            this.refreshInProgress.set(null);
+                            return this.value.updateAndGet(existingValue -> Mono.just(response));
+                        })
+                        .doOnError(throwable -> {
+                            logger.warn("Background refresh task failed", throwable);
+                            this.refreshInProgress.set(null);
+                        })
+                        .cache();
+        }
 
-                }
-                return this.refreshInProgress == null ? valueMono : refreshInProgress;
-            });
+        /***
+         * If there is no refresh in progress background task, then create a new one, else skip
+         *
+         * @param createRefreshFunction the createRefreshFunction
+         * @return if there is already a refreshInProgress task ongoing, then return Mono.empty, else return the newly created background refresh task
+         */
+        public Mono<TValue> refresh(Function<TValue, Mono<TValue>> createRefreshFunction) {
+            if (this.refreshInProgress.compareAndSet(null, this.createBackgroundRefreshTask(createRefreshFunction))) {
+                logger.debug("Started a new background task");
+                return this.refreshInProgress.get();
+            }
+
+            logger.debug("Background refresh task is already in progress, skip creating a new one");
+            return Mono.empty();
```

Comment:
Why are we returning `Mono.empty()`? It doesn't solve any purpose. 
Since we have a null check above on line 137, better to return null and not execute this Mono.empty(). Currently it will consume a thread / CPU for nothing. Thoughts?

Reply:
yea, that also works

Reply:
updated